database/query/profiles.sql

```diff
@@ -60,7 +60,21 @@ WHERE profiles.project_id = $1 AND lower(profiles.name) = lower(sqlc.arg(name));
 
 -- name: ListProfilesByProjectID :many
 SELECT * FROM profiles JOIN entity_profiles ON profiles.id = entity_profiles.profile_id
-WHERE profiles.project_id = $1;
+WHERE profiles.project_id = $1
+AND (
+    -- the most common case first, if the include_labels is empty, we list profiles with no labels
+    -- we use coalesce to handle the case where the include_labels is null
+    (COALESCE(cardinality(sqlc.arg(include_labels)::TEXT[]), 0) = 0 AND profiles.labels = ARRAY[]::TEXT[]) OR
+    -- if the include_labels arg is equal to '*', we list all profiles
+    sqlc.arg(include_labels)::TEXT[] = ARRAY['*'] OR
+    -- if the include_labels arg is not empty and not a wildcard, we list profiles whose labels are a subset of include_labels
+    (COALESCE(cardinality(sqlc.arg(include_labels)::TEXT[]), 0) > 0 AND profiles.labels @> sqlc.arg(include_labels)::TEXT[])
+) AND (
+    -- if the exclude_labels arg is empty, we list all profiles
```

Comment:
For a very low number of profiles it seems to be negligible (compared with the old query the cost went up from `4..9` to `4..10.5`). Let me run queries with a more realistic number of profiles (dozens) and projects.

Are you suggesting that internally (in engine etc) where we don't care about the labels at least not at the moment, we run the query without label support? That seems sensible. But in the handler we'd still default to using the query with label support because the default handler parameters would be to list profiles without labels, right?

Reply:
What about having multiple defined SQL queries for the use-cases we support for filtering? Then we'd have an abstraction that determines which one to use.

Reply:
> Are you suggesting that internally (in engine etc) where we don't care about the labels at least not at the moment, we run the query without label support? That seems sensible. But in the handler we'd still default to using the query with label support because the default handler parameters would be to list profiles without labels, right?

Yes, that was what I was thinking.

Reply:
> > Are you suggesting that internally (in engine etc) where we don't care about the labels at least not at the moment, we run the query without label support? That seems sensible. But in the handler we'd still default to using the query with label support because the default handler parameters would be to list profiles without labels, right?
> 
> Yes, that was what I was thinking.

That's a good idea, I will incorporate that, thanks.

Reply:
I did that. I hit two snags (that's why it took me a bit):
 - I couldn't force sqlc to generate the same return structure for the two SQL queries, even after I've changed them from `SELECT *` to returning a set of named columns
 - to solve that and be able to pass the row around, I could think of 1) using an interface which is what I did or alternatively I could 2) create a wrapper around the function that lists with labels to convert its results into `ListProfilesByProjectIDRow` which seems slightly less elegant, but probably safer.

@JAORMX @dmjb WDYT?

Reply:
> Is there any performance penalty for doing this, versus leaving the original behaviour and having a second dedicated query to filter by label type?

To circle back to this, I created about 5000 profiles in around 2000 projects. Since the queries are still quite smart that they drill down to profiles in the selected project first (we fixed the indexing yesterday), both queries are very similar in performance. The non-filtering one that just does:
```sql
EXPLAIN ANALYSE SELECT * FROM profiles JOIN entity_profiles ON profiles.id = entity_profiles.profile_id
    WHERE profiles.project_id = '3ed3d6d5-42a5-4e00-881f-63fe35c0dfb6';
```
had cost 4.6-62 and took 1.2ms planning and 0.4ms execution. The one using filters:
```sql
EXPLAIN ANALYSE SELECT * FROM profiles JOIN entity_profiles ON profiles.id = entity_profiles.profile_id
WHERE profiles.project_id = '3ed3d6d5-42a5-4e00-881f-63fe35c0dfb6'
  AND (
    -- the most common case first, if the include_labels is empty, we list profiles with no labels
    -- we use coalesce to handle the case where the include_labels is null
    (COALESCE(cardinality(ARRAY['obsolete']::TEXT[]), 0) = 0 AND profiles.labels = ARRAY[]::TEXT[]) OR
        -- if the include_labels arg is equal to '*', we list all profiles
    ARRAY['obsolete']::TEXT[] = ARRAY['*'] OR
        -- if the include_labels arg is not empty and not a wildcard, we list profiles whose labels are a subset of include_labels
    (COALESCE(cardinality(ARRAY['obsolete']::TEXT[]), 0) > 0 AND profiles.labels @> ARRAY['obsolete']::TEXT[])
    ) AND (
    -- if the exclude_labels arg is empty, we list all profiles
    COALESCE(cardinality(ARRAY['stacklok:health']::TEXT[]), 0) = 0 OR
        -- if the exclude_labels arg is not empty, we list profiles whose labels are not a subset of exclude_labels
    NOT profiles.labels @> ARRAY['stacklok:health']::TEXT[]
    );
```
had very surprisingly lower cost as reported by `EXPLAIN` but the times were a slightly bit higher, planning took 1.4ms and execution 0.46ms. All in all, a very small difference.

Reply:
The small difference even makes me question the two SQL queries, but even if the performance difference is not there now, I can see us extending the label queries quite a bit in the future and then it might make sense to have them separately also for performance reasons.

Reply:
> I did that. I hit two snags (that's why it took me a bit):
> 
> * I couldn't force sqlc to generate the same return structure for the two SQL queries, even after I've changed them from `SELECT *` to returning a set of named columns
> * to solve that and be able to pass the row around, I could think of 1) using an interface which is what I did or alternatively I could 2) create a wrapper around the function that lists with labels to convert its results into `ListProfilesByProjectIDRow` which seems slightly less elegant, but probably safer.
> 
> @JAORMX @dmjb WDYT?

I'd like to also check if https://docs.sqlc.dev/en/stable/howto/embedding.html can make the generated code nicer.

Reply:
embedding structs made the code somewhat cleaner and in general I think we should use that going forward for sqlc code, I like that it makes the resulting rows much more reusable.

I also ended up pushing the API part to this patch to make it externally testable. We can still add the separate `with_label`/`without_label` query parameters, but that would be trivial addition to the proto+filing up the query labels.

Reply:
If both queries offer similar performance, I'm happy to take the single query approach.

Reply:
I'd expect the two queries to offer similar performance; my mental model is:

1. Fetch all the rows that match the `profile.project_id`
2. In the case of the label-matching query, drop some of the rows before sending them back over the wire.

My intuition is that the extra filtering logic is probably cheap in comparison with the serialization over the wire, and both queries are reading roughly the same rows out of the index, though the label filters may send back fewer rows.